### PR TITLE
[SPARK-27378][YARN] YARN support for GPU-aware scheduling

### PR DESCRIPTION
**Pull request by tgravescs**
## What changes were proposed in this pull request?

Add yarn support for GPU-aware scheduling. Since SPARK-20327 already added yarn custom resource support, this jira is really just making sure the spark resource configs get mapped into the yarn resource configs and user doesn't specify both yarn and spark config for the known types of resources (gpu and fpga are the known types on yarn).

You can find more details on the design and requirements documented: https://issues.apache.org/jira/browse/SPARK-27376

Note that the running on yarn docs already state to use it, it must be yarn 3.0+. We will add any further documentation under SPARK-20327

## How was this patch tested?

Unit tests and manually testing on yarn cluster

### CODE REVIEW

**Comment by tgravescs** [comment]
need to fix test with hadoop 2.7 compilation

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 1]
```diff
@@ -172,7 +172,7 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
   }
 
   /** Length of time to wait while draining listener events. */
-  val WAIT_TIMEOUT_MILLIS = 10000
+  val WAIT_TIMEOUT_MILLIS = 30000
```
Comment (tgravescs) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: I was seeing some intermittent timeouts on this test so increased, I can back it out if people think this is to long.

**Comment by tgravescs** [comment]
@mengxr @jiangxb1987

**Comment by tgravescs** [comment]
I need to update this based on https://github.com/apache/spark/pull/24615

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 2]
```diff
@@ -39,6 +40,25 @@ object YarnSparkHadoopUtil {
...
+   * spark.yarn.*.resource and the spark.*.resource configs.
+   */
+  private[yarn] def getYarnResourcesFromSparkResources(
+      confPrefix: String,
+      sparkConf: SparkConf
+      ): Map[String, String] = {
+    Map("gpu" -> YARN_GPU_RESOURCE_CONFIG, "fpga" -> YARN_FPGA_RESOURCE_CONFIG).map {
+      case (rName, yarnName) =>
+        val resourceCountSparkConf = s"${confPrefix}${rName}${SPARK_RESOURCE_COUNT_SUFFIX}"
+        (yarnName -> sparkConf.getOption(resourceCountSparkConf).getOrElse("0"))
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: `get(resourceCountSparkConf, "0")`

**Review by vanzin — COMMENTED** [review]
Looks fine. Let's see how far the hadoop-3.2 build gets.

**Comment by vanzin** [comment]
retest this please

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala (hunk 2 above)
Comment (tgravescs) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: updated

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
Just to confirm: does this cover all of SPARK-27361? There's one item ("YARN can pass GPU info to Spark executor.") which doesn't seem to be covered here, although I'm not really sure what that means (or whether just making sure the resource is requested from YARN is enough).

**Comment by tgravescs** [comment]
yes this covers all of  SPARK-27361. That item can be changed, it was the initial investigate item but yarn doesn't tell you anything about GPU's its allocated to you, so its up to you to discovery the information.
The design jira should cover that: https://issues.apache.org/jira/browse/SPARK-27376

I updated the description to say "How the Spark executor discovers GPU's when run on YARN"
Let me know if that doesn't make sense

**Comment by tgravescs** [comment]
the other thing we could do is change this PR to be SPARK-27378 specific since its a subtask under the whole yarn support side

**Comment by vanzin** [comment]
Changing to SPARK-27378 sounds good.

**Comment by vanzin** [comment]
Merging to master.

**Comment by tgravescs** [comment]
thanks @vanzin

**Pull request closed without merging**